src/encoding_transcode.rs

```diff
@@ -0,0 +1,422 @@
+use crate::internal_events::{
+    DecoderBomRemoval, DecoderMalformedReplacement, EncoderUnmappableReplacement,
+};
+use bytes::{Bytes, BytesMut};
+use encoding_rs::{CoderResult, Encoding};
+
+const BUFFER_SIZE: usize = 4096;
+
+// BOM unicode character (U+FEFF) expressed in utf-8
+// http://unicode.org/faq/utf_bom.html#bom4
+const BOM_UTF8: &[u8] = b"\xef\xbb\xbf";
+const BOM_UTF8_LEN: usize = BOM_UTF8.len();
+
+/// Helps transcoding from the specified encoding to utf8
+pub struct Decoder {
+    buffer: [u8; BUFFER_SIZE],
+    output: BytesMut,
+    inner: encoding_rs::Decoder,
+}
+
+impl Decoder {
+    pub fn new(encoding: &'static Encoding) -> Self {
+        Self {
+            buffer: [0; BUFFER_SIZE],
```

Comment:
This buffer does not appear to actually retain any data between calls to `decode_to_utf8`. Could it be a local in that function instead?

Reply:
Moving it to be local in the function will work but this was done to save on the buffer initialization/allocation when `decode_to_utf8` is called repeatedly (like in the file source, which initializes the `Decoder` once and calls `decode_to_utf8` for each line).

Reply:
:+1: That makes sense. You might be able to get an uninitialized buffer with `MaybeUninit`, but probably only by adding `unsafe` blocks.